### ISSUE TITLE
Send trait implementation

### ISSUE THREAD

**Issue by sugar700** [marmeladema/clru-rs#39]
Currently `CLruCache` never implements `Send` trait because `Key<K>` uses `Rc` internally. However in this crate, `Key`/`Rc` is always owned by `CLruCache` and never leaves the structure which means that `CLruCache` could in theory implement `Send` (assuming `K`, `V`, `S` and `W` are all `Send`).

Implementing `Send` would allow for usage of `&Mutex<CLruCache<...>>` from multiple threads.

**Comment by marmeladema** [comment]
Thank you for this. I came to the same conclusion but because I wasn't exactly sure, I never actually implemented anything around this.

I will open a pull request and I'll ask you for your input if that's ok. Do you know any way of detecting any flaws beside manual review? Maybe with Miri? Or thread sanitizer?

**Comment by sugar700** [comment]
Thread sanitizer or Miri could find thread safety issues if an unsound code is ran. For instance, this program will cause Miri error:

```rust
use std::rc::Rc;
use std::thread;

#[derive(Clone)]
struct SendRc<T>(Rc<T>);

unsafe impl<T: Send> Send for SendRc<T> {}

fn main() {
    let s = SendRc(Rc::new(42));
    let sc = s.clone();
    let t = thread::spawn(move || drop(sc));
    drop(s);
    t.join().unwrap();
}
```

That said, the problem with implementing `Send` is that this is a rather weird invariant. `Rc` doesn't implement `Send` because the language cannot ensure that all instances of `Rc` will move to another thread. This is an invariant that pretty much affects entire code, for instance if `CLruCache` were to implement `Clone` trait it would need to be careful to ensure `Rc` won't simply be copied, but instead an entirely new `Rc` would be created with a cloned value in it. Maybe `Key::clone` could be marked as `unsafe` to indicate the invariant.

A safe alternative is to use `Arc` instead, but that introduces potential cost of atomicity in code that doesn't need it.

**Comment by marmeladema** [comment]
Released on 0.5.0
